### ISSUE TITLE
feature-request: remote support

### ISSUE THREAD

**Issue by awwalker** [mfussenegger/nvim-dap#58]
Hi! I am wondering if supporting a 2nd method for attaching to a running debugger process alongside ssh is on your roadmap/something you'd be willing to accept PRs for. 

For a little more context Vscode allows connecting with a `remote` mode which I have used in the past to connect to apps running inside docker containers. The configuration json in `launch.json` looks something like this:
```json
{
      "name": "service attach api_v2",
      "type": "go",
      "request": "attach",
      "mode": "remote",
      "cwd": "${workspaceFolder}/go.git",
      "remotePath": "/opt/gopath/src/github.com/go.git",
      "port": 50029,
}
```

and the shell script that kicks off the debugger on the remote host is something like this:
```sh
#!/bin/bash
set -e
CONTAINER=$1
TARGET_PROCESS=$2
DELVE_PORT=$3
docker exec ${CONTAINER}_1 sh -c "pkill -9 dlv; exit 0"
docker exec ${CONTAINER}_1 pgrep -n -f "$TARGET_PROCESS" | \
  xargs -I "{PID}" docker exec -d ${CONTAINER}_1 bin/delve attach {PID} --headless --listen=:$DELVE_PORT --log --api-version=2
```

I added support for this in vimspector but am trying to migrate to nvim-dap for first class neovim support so would definitely be willing to work on this if you were interested in accepting a PR for it. Thanks!

**Comment by mfussenegger** [comment]
Do you have a link that describes the `remote` mode in more detail? I couldn't find it in the vscode docs after a quick search

It sounds a bit as if it's mostly about executing some action prior to attaching the client to the debug adapter?

**Comment by awwalker** [comment]
Yes I think that's correct. Rather than finding the debugee PID on the localhost or remote machine through ssh we find it using docker commands and then attach to it also using docker commands.

**Comment by awwalker** [comment]
I did some reading @mfussenegger and it seems like the way `attach` is setup is meant to be pretty open and allow all kinds of commands to be run in order to connect to an already running debugger process. From the JSON spec:
```
"The attach request is sent from the client to the debug adapter to attach to a debuggee that is already running.\nSince attaching is debugger/runtime specific, the arguments for this request are not part of this specification."
```
I think vscode allows a tag `preLaunchTask` that I am using to kick of the shell script I pasted above and different debuggers have different ways that they handle the remapping of remotePath with the cwd

**Comment by mfussenegger** [comment]
Yes, `attach` is pretty open. The supported options are up to the individual debug adapter implementations.

With the debug adapter protocol there are usually 4 components involved:

```
DAP-Client  ---- Debug Adapter ---- Debugger --- Debugee
(nvim-dap)                                      (your app)
```


`dap.adapters` (`:help dap-adapters`) is about how to connect or launch a debug adapter.
`dap.configurations` (`:help dap-configurations`) is about how to launch or connect the debugee.


Theoretically there is the option to either run the debug-adapter remote, or to run only the debugger remote. But in the latter case the debug adapter would need to support that functionality.

For example, with debugpy you could run the debug adapter remote:


```
ssh somehost python3 -m debugpy --listen 0.0.0.0:5678 --wait-for-client ./foo.py
```

And create the following entries for nvim-dap:

```lua
dap.adapters.remote_python = function(callback)
  local host = vim.fn.input('Remote IP: ')
  callback({
    type = 'server';
    host = host;
    port = 5678;
  })
end

dap.configurations.python = {
  {
    type = 'remote_python';
    request = 'attach';
    name = "Remote attach";
    port = 5678;
    host = 'localhost';
  },
}
```

The `remote_python` adapter function could also be extended to connect via SSH and launch the debug adapter itself. In your case you could have a `remote_go` adapter that connects via SSH and calls the script.


> I think vscode allows a tag preLaunchTask that I am using to kick of the shell script I pasted above 

Could you share the task?
I'm not sure if it is a good idea to also integrate `tasks.json` support into nvim-dap. That might be worth a separate plugin.

**Comment by awwalker** [comment]
hm yeah I can share the whole thing: 
```json
{
"Version": "2.0.0",
"tasks": [
    {
      "label": "start_delve_consumer_portal",
      "type": "shell",
      "command": "${workspaceFolder}/.vscode/start_delve.sh",
      "args": [
        "consumer_portal",
        "megabin",
        "9005"
      ],
      "presentation": {
        "reveal": "silent"
      }
    }
]
}
```

and then start_delve.sh runs:
```sh
#!/bin/bash

set -e

CONTAINER=$1
TARGET_PROCESS=$2
DELVE_PORT=$3

docker exec devenv_${CONTAINER}_1 sh -c "pkill -9 dlv; exit 0"
docker exec devenv_${CONTAINER}_1 pgrep -n -f "$TARGET_PROCESS" | \
  xargs -I "{PID}" docker exec -d devenv_${CONTAINER}_1 bin/$(cat .thirdparty_version)/delve attach {PID} --headless --listen=:$DELVE_PORT --log --api-version=2
```
and launch.json will have an entry like:
```json
    {
      "name": "service attach consumer_portal",
      "type": "go",
      "request": "attach",
      "mode": "remote",
      "preLaunchTask": "start_delve_consumer_portal",
      "cwd": "${workspaceFolder}/go.git",
      "remotePath": "/opt/gopath/src/github.plaid.com/plaid/go.git",
      "port": 9005,
      "presentation": {
        "group": "attach service"
      },
      "dlvLoadConfig": {
        "maxStringLen": 500,
        "maxArrayValues": 500,
        "maxStructFields": -1
      },
      "internalConsoleOptions": "openOnSessionStart"
    },
```

I agree about not making `task.json` or `preLaunchTask` part of the responsibility of nvim-dap. That's definitely overkill. Im mostly concerned with having to ssh into running docker containers rather than being able to tell my debugger to attach using `docker exec` or `docker attach` since setting up ssh on a container is a little bit of a hassle and the docker commands for it come for free. I think what I'm asking for is something like:
```json
dap.adapters.container_python = function(callback)
  local container = vim.fn.input('Container: ')
  callback({
    type = 'container';
    name = container; // This would be the name of the container running ptvsd or debugpy on port 5678.
    port = 5678;
  })
end
dap.configurations.python = {
  {
    type = 'docker_python';
    request = 'attach';
    name = 'Docker attach';
    container = 'python_container';
    port = 5678;
  },
}
```
Sorry if the switch from go to python is confusing. My main issue with the current setup is the need to have to SSH into the container to start/connect/attach to the debugger process. I would like to be able to through nvim-dap do all of those things that we can do now through SSH with `docker exec` or `docker attach` commands. Does that help clarify things?

**Comment by mfussenegger** [comment]
> Im mostly concerned with having to ssh into running docker containers 

Could you elaborate where ssh comes into play in your example?
As far as I can tell `${workspaceFolder}/.vscode/start_delve.sh` is on your machine, and the docker commands are invoked on your machine?

**Comment by awwalker** [comment]
Yes I think that is true but couldn't the same be said for the ssh commands in the ssh examples? Those get run on the local machine so that the other instructions can get run on the other one. Or am I misunderstanding how the SSH setup works?

**Comment by mfussenegger** [comment]
> Yes I think that is true but couldn't the same be said for the ssh commands in the ssh examples? Those get run on the local machine so that the other instructions can get run on the other one. Or am I misunderstanding how the SSH setup works?


You're right. It was mostly for me to understand what the `remote` mode is supposed to do.
Given that the adapter definitions are currently externalized and not part of `nvim-dap`, I think once https://github.com/mfussenegger/nvim-dap/pull/64 is merged you could define an adapter that does what you want:


```lua
dap.adapters.go = function(cb, config)
  if config.request == 'attach' then
    if config.mode == 'remote' then
      local container = config.container
      local port = config.port or 9005

      vim.fn.system({'path/to/your/script', container, port})
      -- maybe this requires the executable to launch vscode-go
      cb({
        type = 'server',
        host = '127.0.0.1',
        port = port
      })
    else
      -- ...
    end
  else
    -- ...
  end
end
```

You basically extract the items from the configuration that you need to pass on to your script, and then you either launch the debug adapter or you connect to one - depending on what your script already did.

**Comment by mfussenegger** [comment]
@awwalker Did the example help you with your use-case, or is there still something missing in nvim-dap to make your use-case work?

**Comment by awwalker** [comment]
@mfussenegger sorry! I haven't been able to test yet but I'll try to get to it by eow. Thanks for checking in

**Comment by awwalker** [comment]
Ok took some time and got a working non-docker python debugger setup will try with docker this weekend and post back the updates.

**Comment by awwalker** [comment]
ok @mfussenegger sorry for the long silence. I think my limited lua experience is holding me back here. In the example you have posted above is the `config` you are referencing coming from `dap.configurations.go`?

**Comment by mfussenegger** [comment]
> ok @mfussenegger sorry for the long silence. I think my limited lua experience is holding me back here. In the example you have posted above is the `config` you are referencing coming from `dap.configurations.go`?

Yes, the `config` is the configuration from `dap.configurations.go` which you selected - for example when using the `require('dap').continue()` function.

> 
> And where does `cb` come from?

It's a callback function that needs to be called with the final adapter. See `:help dap-adapter`. There is another example that explains the reasoning a bit more.

**Comment by awwalker** [comment]
Ok @mfussenegger I think I'm really close but for some reason my breakpoints aren't getting triggered. I wonder if you can see the problem or give me some ideas as to how to diagnose.

I have in `.config/nvim/lua/debuggers.lua`:
```lua
local M = {}
local goAttachConfig
local localPythonAttachConfig

M.attach_go_debugger = function(args)
    local dap = require "dap"
    local host = "127.0.0.1"
    local port = tonumber(args[2])
    if args and #args > 0 then
        goAttachConfig = {
            type = "go";
            request = "attach";
            mode = "remote";
            dlvToolPath = os.getenv('HOME') .. "/go/bin/dlv";
            remotePath = "/opt/gopath/src/github.plaid.com/plaid/go.git";

            name = args[1];
            port = port;
            -- cwd = vim.fn.cwd();
        }
    end
    if not goAttachConfig then
        print("No remoteGoConfig set! Use :DebugGo <app name> <port> <devenv type>")
        return
    end
    if #args >= 3 and args[3] == "remote" then
        host = "awalker.devenv.plaid.io"
    end
    goAttachConfig["host"] = host
    print("Connecting to: " .. goAttachConfig.host .. ":" .. goAttachConfig.port)
    local out = vim.fn.system({"/Users/awalker/plaid/go.git/scripts/ensure_debugger_session.sh", args[1]})
    local err = vim.v.shell_error
    print(err)
    print(out)
    local session = dap.attach(host, port, goAttachConfig)
    -- dap.repl.open()
end

```

and then in `.config/nvim/init.vim`:
```vim
command! -complete=file -nargs=* DebugGo lua require"debuggers".attach_go_debugger({<f-args>})

```

and when I call `:DebugGo <app_name> <port> localhost` to connect to a container I can see the script is correctly spinning up a delve instance
```sh
Connecting to: 127.0.0.1:50799
^[[36mINFO^[[0m[0000] preparing service for debugging: secmaster
^[[36mINFO^[[0m[0000] Running in local shell
docker exec devenv_secmaster_1 scripts/delve-attach.sh
^[[36mINFO^[[0m[0001] service is prepared for debugging: secmaster
```

and I can prove that this delve instance is ready by running:
```sh
delve connect localhost:50799
```
and seeing the `dlv` prompt come up.

However if I set some breakpoints inside my service code using `nvim-dap` and then trigger the route that should be broken on they are not triggered. Sorry this has taken me so long to get to and that this is extremely specific output but I wonder if I'm just using nvim-dap incorrectly or if there is an obvious error somewhere

**Comment by mfussenegger** [comment]
Is the delve port properly exposed to the host? You could try using `nc` or `telnet` to see if the host/port can be accessed from outside the container.

Could be that nvim-dap can't connect to the delve service because it is being isolated within the container.

That you only see the `initialize` command in the log and no follow up response is an indication that the communication doesn't work.

**Comment by awwalker** [comment]
Hm I think I can `telnet` across my local host to the Docker container on that port:
```telnet localhost 50799                                                                                                                                         
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
```
I actually think I haven't set things up correctly and this has to do with the vscode-go adapter maybe? In your example there is a comment about that. 

How can I be sure the adapter is running? Do I need to start the adapter in the same container delve is running in?

**Comment by awwalker** [comment]
Ok so the code now looks like this:
```lua
M.attach_go_debugger = function(args)
    local dap = require "dap"
    local host = "127.0.0.1"
    local port = tonumber(args[2])
    if args and #args > 0 then
        goAttachConfig = {
            type = "go";
            request = "attach";
            mode = "remote";
            name = "Remote Debugger";
            dlvToolPath = os.getenv('HOME') .. "/go/bin/dlv";
            remotePath = "/opt/gopath/src/github.plaid.com/plaid/go.git";
            port = port;
            cwd = vim.fn.getcwd();
            trace = true;
        }
        goAttachAdapter = {
            type = "executable";
            command = "node";
            args = {os.getenv("HOME") .. "/vscode-go/dist/debugAdapter.js"};
        }
    end
    if not goAttachConfig then
        print("No remoteGoConfig set! Use :DebugGo <app name> <port> <devenv type>")
        return
    end
    if #args >= 3 and args[3] == "remote" then
        host = "awalker.devenv.plaid.io"
    end
    print("Connecting to: " .. host .. ":" .. goAttachConfig.port)
    vim.fn.system({"/Users/awalker/plaid/go.git/scripts/ensure_debugger_session.sh", args[1]})
    local session = dap.launch(goAttachAdapter, goAttachConfig)
end
```
which produces the following logs in `nvim-dap`:
```sh
Every 2.0s: cat vim-dap.log                                                                                                                                                                        awalker-mbp: Wed Dec 23 11:00:27 2020

[ DEBUG ] 2020-12-23T10:58:10Z-0500 ] ...s/awalker/.local/share/nvim/plugged/nvim-dap/lua/dap.lua:1010 ]        "request"       {
  arguments = {
    adapterID = "nvim-dap",
    clientId = "neovim",
    clientname = "neovim",
    columnsStartAt1 = true,
    linesStartAt1 = true,
    locale = "en_US.UTF-8",
    pathFormat = "path",
    supportsRunInTerminalRequest = true
  },
  command = "initialize",
  seq = 0,
  type = "request"
}
[ ERROR ] 2020-12-23T10:58:10Z-0500 ] ...s/awalker/.local/share/nvim/plugged/nvim-dap/lua/dap.lua:986 ] "stderr"        {
  args = { "/Users/awalker/vscode-go/dist/debugAdapter.js" },
  command = "node",
  type = "executable"
}       "(node:29631) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.\n(Use `node --trace-deprecation
...` to show where the warning was created)\n"
[ DEBUG ] 2020-12-23T10:58:15Z-0500 ] ...s/awalker/.local/share/nvim/plugged/nvim-dap/lua/dap.lua:840 ] {
  body = {
    supportsConditionalBreakpoints = true,
    supportsConfigurationDoneRequest = true,
    supportsSetVariable = true
  },
  command = "initialize",
  request_seq = 0,
  seq = 1,
  success = true,
  type = "response"
}
[ DEBUG ] 2020-12-23T10:58:15Z-0500 ] ...s/awalker/.local/share/nvim/plugged/nvim-dap/lua/dap.lua:1010 ]        "request"       {
  arguments = {
    cwd = "/Users/awalker/plaid/go.git",
    dlvToolPath = "/Users/awalker/go/bin/dlv",
    mode = "remote",
    name = "Remote Debugger",
    port = 50799,
    remotePath = "/opt/gopath/src/github.plaid.com/plaid/go.git",
    request = "attach",
    trace = true,
    type = "go"
  },
  command = "attach",
  seq = 2,
  type = "request"
}
[ DEBUG ] 2020-12-23T10:58:15Z-0500 ] ...s/awalker/.local/share/nvim/plugged/nvim-dap/lua/dap.lua:840 ] {
  event = "initialized",
  seq = 2,
  type = "event"
}
[ DEBUG ] 2020-12-23T10:58:15Z-0500 ] ...s/awalker/.local/share/nvim/plugged/nvim-dap/lua/dap.lua:1010 ]        "request"       {
  command = "configurationDone",
  seq = 3,
  type = "request"
}
[ DEBUG ] 2020-12-23T10:58:15Z-0500 ] ...s/awalker/.local/share/nvim/plugged/nvim-dap/lua/dap.lua:840 ] {
  command = "attach",
  request_seq = 2,
  seq = 3,
  success = true,
  type = "response"
}
[ DEBUG ] 2020-12-23T10:58:15Z-0500 ] ...s/awalker/.local/share/nvim/plugged/nvim-dap/lua/dap.lua:840 ] {
  command = "configurationDone",
  request_seq = 3,
  seq = 4,
  success = true,
  type = "response"
}
```

and the following logs from `delve`:
```sh
API server listening at: [::]:50799
2020-12-23T15:58:10Z debug layer=rpc API server pid = 361
2020-12-23T15:58:12Z debug layer=rpc <- RPCServer.SetApiVersion(api.SetAPIVersionIn{"APIVersion":2})
2020-12-23T15:58:12Z debug layer=rpc -> *api.SetAPIVersionOut{} error: ""
2020-12-23T15:58:12Z debug layer=rpc (async 2) <- RPCServer.Command(api.DebuggerCommand{"name":"continue","ReturnInfoLoadConfig":null})
2020-12-23T15:58:15Z debug layer=rpc <- RPCServer.GetVersion(api.GetVersionIn{})
2020-12-23T15:58:15Z debug layer=rpc -> *api.GetVersionOut{"DelveVersion":"Version: 1.5.1\nBuild: $Id: bca418ea7ae2a4dcda985e623625da727d4525b5 $","APIVersion":2,"Backend":"native","TargetGoVersion":"","MinSupportedVersionOfGo":"1.13.0","MaxSupportedVersionOfGo":"1.15.0"} error: ""
2020-12-23T15:58:15Z debug layer=rpc (async 2) <- RPCServer.State(rpc2.StateIn{"NonBlocking":true})
2020-12-23T15:58:15Z debug layer=rpc (async 2) -> rpc2.StateOut{"State":{"Running":true,"Recording":false,"Threads":null,"NextInProgress":false,"exited":false,"exitStatus":0,"When":""}} error: ""
```
And everything seems to work! Thanks so much I'll clean up the code and make it conform more the API and we can close this issue and if you want I can submit a PR to document this use case more

**Comment by mfussenegger** [comment]
> if you want I can submit a PR to document this use case more

You could also extend the Wiki. It can be edited by anyone.

**Comment by awwalker** [comment]
> You could also extend the Wiki. It can be edited by anyone.

I can either add a new page to the wiki for `Remote Debugging` and fill it with examples as I add them for my own use cases (Python and Go) or I can just add a section under the current examples in the existing `Debug Adapter installation` section. Which would you prefer?

**Comment by mfussenegger** [comment]
>  Which would you prefer?

I've a slight preference for a new page "Remote Debugging"

**Comment by awwalker** [comment]
I'm not sure if theres a better way to share what I wrote up so I'm going to just publish the new page and we can discuss edits to it in a new issue? Is that ok @mfussenegger? I also tweaked the title to "Local and Remote Debugging with Docker" since that was the specific use case I created configurations for, but it goes over:

```md
There are some basic steps that remain the same across both remote and local Docker Debugging:
1. Expose a port in your Docker container responsible for debugging.
2. Modify your containers [security settings](#Docker-Security-Settings) to allow ptrace to run correctly.
3. Install the program/package responsible for starting the debug server (and maybe adapter) in your container.
4. Install the program/package responsible for launching the debug adapter locally (if necessary).
5. Start/Attach the debug server (and potentially adapter) inside your container to your code.
6. Launch the debug adapter locally (if not done by the debug server in the container).
7. Connect nvim-dap to either the debug adapter you spawned locally or the adapter spun up by your debug server in the container.
```
For both debugpy and delve

**Comment by mfussenegger** [comment]
> I'm not sure if theres a better way to share what I wrote up so I'm going to just publish the new page and we can discuss edits to it in a new issue? Is that ok @mfussenegger? 

Sure, sounds great :+1:

**Comment by davidvalaz** [comment]
Hey @awwalker - Could you also share your Dockerfile, docker-compose.yml and the specific commands you ran with Python? Config seems great in the Wiki (and your Dotfiles) but seemingly having some connection issues. Will be happy to own this and write up in the Wiki once solved

**Comment by awwalker** [comment]
Hey @davidvalaz what are the errors you're seeing? I can dig up the stuff I wrote for Python

**Comment by awwalker** [comment]
@davidvalaz can you look at https://github.com/awwalker/nvim-dap/pull/1 and see if this helps you?

**Comment by davidvalaz** [comment]
Hey guys sorry for the silence on this. I managed to get it working in the end. @awwalker your config is super clear and well thought out!

I'm going to have a play with Pytest today as my workflow is TDD based. Saw someone posted a Docker command over on `nvim-dap-python` so will look at that.
